src/landlab/components/soil_grading/soil_grading.py

```diff
@@ -0,0 +1,557 @@
+             is the percentage of volume that remains in the parent size fraction
+             after fragmentation, BBB and CCC (and so on) are the proportion of
+             daugther particles in the next (smaller) size classes.
+         grain_max_size : float (m)
+             The maximal grain size represented in the grading distribution
+         power_of : float (-)
+             The power relation between parent and dauther grain volume
+         n_sizes : int (-)
+             The number of classes represented in the distribution
+         precent_of_volume_in_spread : float (-)
+             The percentage of mass from parent spread between the daughter grains
+              in case of 'spread' flag is found in the grading name.
+        """
+
+        # Make sure that grading_name is valid
+        grading_name_split = grading_name.split("-")
+        is_p = grading_name_split[0][0] == "p"
+        is_size = np.size(grading_name_split) > 2
+        is_N = len(grading_name_split[0]) == 2 and grading_name_split[0][1].isnumeric()
+        is_numeric = True
+        sum_values = 0
+        for s in grading_name.split("-")[1:]:
+            if s != "spread":
+                is_numeric *= s.isnumeric()
+                if is_numeric:
+                    sum_values += int(s)
+        is_smaller_than_100 = sum_values <= 100 or "spread" in grading_name
+        if is_p * is_size * is_numeric * is_smaller_than_100 * is_N is False:
+            raise ValueError("grading name provided not valid")
+
+        # Create grading distribution based on the grading name
+        N = int(grading_name_split[0][1])
+        parent = np.copy(grain_max_size)
+        limits = np.zeros((n_sizes, 2))
+        limits[0, 1] = grain_max_size
+        for cnt in range(0, n_sizes):
+            limits[cnt, 1] = parent
+            parent = parent * (1 / N) ** (power_of)
+            limits[cnt, 0] = parent
+        limits = limits[::-1]  # sort in ascend order
+        meansizes = np.asarray((limits[:, 0] + limits[:, 1]) / 2)
+
+        # Create fragmentation vector describing the fragmentation pattern
+        precents = np.array(
+            [
+                float(s)
+                for s in grading_name.split("-")
+                if s.replace(".", "", 1).isdigit()
+            ]
+        )
+        if "spread" in grading_name:
+            precents_to_add = np.ones((1, n_sizes)) * precent_of_volume_in_spread
+            precents = np.append(precents, precents_to_add)
+        fragmentation_pattern = precents / 100
+
+        return meansizes, limits, fragmentation_pattern
+
+    def check_fragmentation_pattern(self):
+        is_length = len(self._fragmentation_pattern) >= 2 * (
+            len(self._fragmentation_pattern) <= len(self._meansizes)
+        )
+        is_1 = np.sum(self._fragmentation_pattern) <= 1
+        if not is_length * is_1 is True:
+            raise ValueError("fragmentation pattern provided not valid")
```

Comment:
@yuvalsmilo Could you clean this up? I find it hard to follow. Perhaps is should be the following,
```python
if (
    len(self._fragmentation_pattern) < 2
    or len(self._fragmentation_pattern) > len(self._meansizes)
    or np.sum(self._fragmentation_patter) > 1.0
):
    raise ValueError("fragmentation pattern provided not valid")
```

Reply:
OK thanks, changed